File: app/channels/chat_channel.rb
# app/channels/chat_channel.rb
class ChatChannel < ApplicationCable::Channel
  def subscribed
    stream_from "chat_channel"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def speak(data)
    message = Message.create!(content: data['message'], user: User.find(data['user']))

    # ActionCable.server.broadcast "chat_channel", message: render_message(message)

    ActionCable.server.broadcast("chat_channel", {
      message: "<p><strong>#{message.user.email_address}:</strong> #{message.content}</p>"
    })
  end

  private

  def render_message(message)
    ApplicationController.renderer.render(partial: 'messages/message', locals: { message: message })
  end
end

File: app/javascript/channels/chat_channel.js
import consumer from "channels/consumer";

consumer.subscriptions.create("ChatChannel", {
  connected() {
    console.log("Connected to the chat channel!");
  },

  disconnected() {
    console.log("Disconnected from the chat channel.");
  },

  received(data) {
    const messages = document.getElementById("messages");
    messages.insertAdjacentHTML("beforeend", data.message);
  },

  speak(message, user) {
    return this.perform("speak", { message: message, user: user });
  }
});

document.addEventListener("DOMContentLoaded", function () {
  const input = document.getElementById("chat-input");
  const button = document.getElementById("send-button");
  const inputUser = document.getElementById("user-chat-input"); // User input field

  button.addEventListener("click", function () {
    const message = input.value;
    const user = inputUser ? inputUser.value : "Anonymous"; // Fallback if no user input

    if (message) {
      consumer.subscriptions.subscriptions[0].speak(message, user);
      input.value = "";
    }
  });
});
